0000-notes-on-oer-encoding/0000-notes-on-oer-encoding.md

```diff
@@ -0,0 +1,285 @@
+
+Nevertheless, designers of new Interledger protocols SHOULD consider staying within the boundaries set by this document. Deviations from these norms should be carefully considered. This is intended to minimize the load on implementors who should be able to write components such as date encoding and decoding once and reuse it across protocols.
+
+# How to use this document
+
+This document is intended as a first reference for implementors to use when writing code that serializes and deserializes Interledger formats. We intend to make this document sufficient on its own, such that reading the official ASN.1/OER specifications is not required to correctly implement Interledger formats.
+
+Implementors SHOULD NOT aim to create a full-featured OER parser, since Interledger intentionally does not use some of the more complex features of ASN.1. Implementors MAY use a full-featured CANONICAL-OER implementation, but we expect most will prefer a lighter-weight implementation that only handles types that are actually used in the Interledger protocols.
+
+# Foundation
+
+Interledger encodings are based on and compliant with Abstract Syntax Notation One (ASN.1) using Canonical Octet Encoding Rules (CANONICAL-OER).
+
+## ASN.1
+
+### What is ASN.1?
+
+Abstract Syntax Notation One (ASN.1) is an international standard for describing abstract data types. ITU gives the following definition:
+
+> Abstract Syntax Notation number One
+> is a standard that defines a formalism
+> for the specification of abstract data types.
+
+In plain English, ASN.1 describes what your data can be, without making any statements about how it could be encoded in bits and bytes.
+
+### Why ASN.1?
+
+A protocol like Interledger is fundamentally a set of guidelines for a conversation between different computers. The more precise this set of guidelines is, the less likely it is that there will be a misunderstanding. It is the job of the protocol designer to specify this conversation with as little ambiguity as possible.
+
+Using a formalized language like ASN.1 to describe one's protocol means that one can use automated tools to verify the consistency and completeness of one's specification. This is a very valuable tool. It also means that one can refer to existing specifications which already cover many edge cases. Otherwise, each protocol would have to answer all possible edge cases itself, which quickly becomes tedious.
+
+ASN.1 is by far the most widely used formalized language for defining protocols. As an ISO and ITU standard, it is used in telecommunications protocols, cryptography standards and many more.
+
+The Interledger team chose ASN.1 because of it's long history and robust ecosystem. It may not be the most modern of standards, but it is extremely powerful and comes with huge amounts of implementation experience. Also, there exist high-quality tools for protocol designers.
+
+## CANONICAL-OER
+
+### What is OER/CANONICAL-OER?
```

Comment:
What I'm missing from the intro up to here is how awesome it is that the only special thing you really need to implement to support this format is for variable-length fields. Everything else is encoded in a way that most programming languages support out of the box. I'm not sure you could say that for any other binary format